### ISSUE TITLE
concatWithValues produce sometimes two same values instead of one or out of order under heavy load

### ISSUE THREAD

**Issue by koldat** [reactor/reactor-core#2567]
Our code that runs in multithread and is asynchronously encoding output json sometimes randomly fails as values produces from concatWithValues is returned twice. It also happens sometimes that these are out of order. I have extended out encoder to assert order and content of produced elements. 
 
## Expected Behavior
concatWithValues with one values should append value only once.

## Actual Behavior
concatWithValues return one value two times sometime or out of order.

## Steps to Reproduce
Unfortunately I was not able to do the same in isolated unit test, but here is source code I have. I am executing exactly same query thousand times per second with 100 concurrency. It fails in very rare situations (per 50,000 queries)

```java
        // Index for asserting
        AtomicInteger index = new AtomicInteger();

        Mono<Query> authorized = authorizeAsync(session, query)
            .publishOn(readScheduler);
        Flux<Object> resultFlux = authorized.flatMap(q -> {
            return Mono.just(queryService.query(session, q, request));
        })
        .flatMapMany(graph -> {
            Flux<Object> result = Flux.<Object>just(
                enc.writeStartObject(),
                enc.writeArrayFieldStart("vertices"))
            .concatWith(Flux.fromIterable(graph.getVertices() == null ? Collections.emptyList() : graph.getVertices()))
            .concatWithValues(
                enc.writeEndArray(),
                enc.writeArrayFieldStart("edges"))
            .concatWith(Flux.fromIterable(graph.getEdges() == null ? Collections.emptyList() : graph.getEdges()))
            .concatWithValues(enc.writeEndArray()) // This is the problematic line
            .concatWithValues(enc.writeEndObject());

            return result;
        })
       // Annotate every item with index
        .map(v -> new Pair<>(index.incrementAndGet(), v));
        
        // transform to json buffers and assert order
        enc.apply(resultFlux, () -> createWriter(view, projectionFilter))
            .subscribeOn(readScheduler);
```
Output:

```
2021-01-22 13:13:23.701  WARN 2528 --- [    tas.read-15] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=31, second=ForkEdgeItem@2ffe7ad5]
2021-01-22 13:13:23.701  WARN 2528 --- [    tas.read-15] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=32, second=ForkEdgeItem@3ef397f3]
2021-01-22 13:13:23.701  WARN 2528 --- [    tas.read-15] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=33, second=ForkEdgeItem@4a461f07]
2021-01-22 13:13:23.702  WARN 2528 --- [    tas.read-15] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=34, second=JsonAsStreamEncoder$$Lambda$3752/0x0000000100e6e440@2cd8ce4]

Sometimes same value on index 35 as in 34 in very rare cases (enc.writeEndArray()) from  "problematic line" above
2021-01-22 13:13:23.702  WARN 2528 --- [    tas.read-15] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=35, second=JsonAsStreamEncoder$$Lambda$3752/0x0000000100e6e440@2cd8ce4]


Or Out of order
2021-01-22 13:51:24.456  WARN 4049 --- [    tas.read-26] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=33, second=ForkEdgeItem@8f2e185]
2021-01-22 13:51:24.457  WARN 4049 --- [    tas.read-26] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=34, second=ForkEdgeItem@24e7d361]
2021-01-22 13:51:24.457  WARN 4049 --- [    tas.read-26] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: Pair [first=35, second=JsonAsStreamEncoder$$Lambda$3390/0x0000000100dcc040@1817bcdd]
2021-01-22 13:51:24.457  WARN 4049 --- [    tas.read-26] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: 

Here it jumped to 37
Pair [first=37, second=JsonAsStreamEncoder$$Lambda$3396/0x0000000100dcdc40@1ff557db]
2021-01-22 13:51:24.457  WARN 4049 --- [    tas.read-26] c.c.a.c.jsonstream.JsonAsStreamEncoder   : JsonAsStremEncoder last[]: 

And back to 36. Notice that value is same as for index 35 (enc.writeEndArray()) from  "problematic line" above
Pair [first=36, second=JsonAsStreamEncoder$$Lambda$3390/0x0000000100dcc040@1817bcdd]
```

Similar issue is in other sources. All of these are using more concatWithValues in a row. Strange is that if it fails then always on that index (item with enc.writeEndArray())

## Possible Solution
I tried to put volatile or atomic operations into FluxConcatArray index and produced fields hoping other thread did not refreshed memory, but did not help. My workaround is to use concatMap with Multiple fluxes, or even concatWithValues with two values works. It is very strange.

I tried Executor scheduler as well as bounded elastic.

## Your Environment
AdoptOpenJDK (build 11.0.9.1+1)
Spring Boot 2.3.7.RELEASE
reactor-core-3.3.12.RELEASE
reactor-netty-0.9.15.RELEASE
CentOS docker image

**Comment by koldat** [comment]
I tried to simulate the problem and squeezed operators to absolute minimum to get an exception that is related to threading. It is little bit different symptom, but should not definitely happen. I get an exception under one second now:

```java
public class FFTest {
    private static final Logger log = LoggerFactory.getLogger(FFTest.class);

    @Test
    public void test() {
        int round = 0;
        try {
            while (true) {
                Flux.range(0, 10)
                    .concatWithValues(10, 11, 12, 13)
                    .concatWith(Flux.range(14, 100 - 14))
                    .limitRate(16, 2)
                    .publishOn(Schedulers.boundedElastic(), 16)
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockLast();
                log.info("Round : " + (round++));
            }
        } catch(Exception e) {
            log.error("", e);
        }
    }
}
```

Output:

```
13:06:20.429 [main] INFO com.ca.apm.tas.FFTest - Round : 0
......
13:06:20.442 [main] INFO com.ca.apm.tas.FFTest - Round : 38
13:06:20.442 [main] INFO com.ca.apm.tas.FFTest - Round : 39
13:06:20.450 [main] ERROR com.ca.apm.tas.FFTest - 
reactor.core.Exceptions$OverflowException: Queue is full: Reactive Streams source doesn't respect backpressure
	at reactor.core.Exceptions.failWithOverflow(Exceptions.java:234)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.onNext(FluxPublishOn.java:232)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:183)
	at reactor.core.publisher.FluxRange$RangeSubscription.slowPath(FluxRange.java:154)
	at reactor.core.publisher.FluxRange$RangeSubscription.request(FluxRange.java:109)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.drainLoop(Operators.java:2251)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.drain(Operators.java:2178)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:2124)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.poll(FluxPublishOn.java:642)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runAsync(FluxPublishOn.java:411)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:526)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
```


This test is working fine when:

 - comment limitRate(16, 2)
 - comment publishOn(Schedulers.boundedElastic(), 16)
 - Use Flux.range(0, 100) instead of Flux.range(0, 10).concatWithValues(10, 11, 12, 13).concatWith(Flux.range(14, 100 - 14))

**Comment by koldat** [comment]
I added custom logger as with just log it does not happen (slowing down and issue is not there). It looks to me that under some conditions publishOn switches thread and it mess up the state. I think it should not be possible as publishOn is using Worker that is bounded to single thread pool (boundedElastic), but it happens:

```java
	      	Flux.range(0,10)
                .publishOn(reader)
                .concatWithValues(10, 11, 12, 13)
                .concatWith(Flux.range(14, 100-14))
                .log() // Logger 1
                .publishOn(reader, 16)
                .log() // Logger 2
                .subscribeOn(reader)
                .blockLast();
```

And now See line 31. log is simply adding event to synchronized list so order is correct.

```Index: [Thread] - Name | event
0: [tas-reader-1] - reactor.Flux.OnAssembly.339 | onSubscribe([Fuseable] FluxOnAssembly.OnAssemblySubscriber)
1: [tas-reader-1] - reactor.Flux.OnAssembly.340 | onSubscribe([Fuseable] FluxOnAssembly.OnAssemblySubscriber)
2: [tas-reader-1] - reactor.Flux.OnAssembly.340 | request(unbounded)
3: [tas-reader-1] - reactor.Flux.OnAssembly.339 | request(16)
4: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(0)
5: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(1)
6: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(2)
7: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(3)
8: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(4)
9: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(5)
10: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(6)
11: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(7)
12: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(8)
13: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(9)

14: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(0)
15: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(1)
16: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(2)
17: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(3)
18: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(4)
19: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(5)
20: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(6)
21: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(7)
22: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(8)

23: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(10)
24: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(11)

25: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(9)

26: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(12)

27: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(10)
28: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(11)

29: [tas-reader-2] - reactor.Flux.OnAssembly.339 | onNext(13)
30: [tas-reader-3] - reactor.Flux.OnAssembly.339 | request(12)
31: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(14) <-- Here it switches thread
32: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(15)
33: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(16)
34: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(17)
35: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(18)
36: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(19)
37: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(20)
38: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(21)
39: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(22)
40: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(23)
41: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(24)
42: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(25)
43: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(26)
44: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(27)
45: [tas-reader-3] - reactor.Flux.OnAssembly.339 | onNext(28)
46: [tas-reader-3] - reactor.Flux.OnAssembly.339 | cancel()

47: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(12)
48: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(13)
49: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(14)
50: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(15)
51: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(16)
52: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(17)
53: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(18)
54: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(19)
55: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(20)
56: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(21)
57: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(22)
58: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(23)

59: [tas-reader-3] - reactor.Flux.OnAssembly.339 | request(12)

60: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(24)
61: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(25)
62: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(26)
63: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onNext(27)
64: [tas-reader-3] - reactor.Flux.OnAssembly.340 | onError(reactor.core.Exceptions$OverflowException: Queue is full: Reactive Streams source doesn't respect backpressure)
65: [tas-reader-3] - reactor.Flux.OnAssembly.340 
```

**Comment by koldat** [comment]
@OlegDokuka I found the issue for my unit test. It was really possible to emit more values than requested, And possibly can cause the same in rsocket issue I was referencing in case server does not follow backpressure. Can you please review?

**Comment by koldat** [comment]
I will try to build custom jar and test original problem as well in our application.

**Comment by koldat** [comment]
I have found another simple test that actually finally simulates duplicate value and cause again overflow:

```java
    private static Scheduler reader = Schedulers.newBoundedElastic(32, Integer.MAX_VALUE, "tas-reader");
    
    public static void main(String[] args) throws InterruptedException {
        int round = 0;
        while (true) {
        	Mono.just(0)
        	.flatMapMany(i -> {
        		return Flux.range(0,10)
	                .publishOn(reader)
	                .concatWithValues(10)
	                .concatWithValues(11)
	                .concatWithValues(12)
	                .concatWithValues(13)
	                .concatWithValues(14)
	                .concatWithValues(15)
	                .concatWithValues(16)
	                .concatWithValues(17)
	                .concatWith(Flux.range(18, 100-18));
        	})
            //.log()
            .publishOn(reader, 16)
            //.log()
            .subscribeOn(reader)
            .blockLast();

        	System.out.println("Round : " + (round++));
        }
    }
```

Output
```
.....
25: [tas-reader-2] - reactor.Flux.PublishOn.81446 | onNext(8)
26: [tas-reader-2] - reactor.Flux.PublishOn.81446 | onNext(9)
27: [tas-reader-2] - reactor.Flux.PublishOn.81446 | onNext(10)
28: [tas-reader-2] - reactor.Flux.PublishOn.81446 | onNext(11)
29: [tas-reader-1] - reactor.Flux.MonoFlatMapMany.81445 onNext(13)
30: [tas-reader-2] - reactor.Flux.MonoFlatMapMany.81445 request(12)
31: [tas-reader-1] - reactor.Flux.MonoFlatMapMany.81445 onNext(14) <-- Here is duplicate
32: [tas-reader-2] - reactor.Flux.MonoFlatMapMany.81445 onNext(14) <-- Here is duplicate
33: [tas-reader-1] - reactor.Flux.MonoFlatMapMany.81445 onNext(15)
34: [tas-reader-1] - reactor.Flux.MonoFlatMapMany.81445 onNext(16)
35: [tas-reader-1] - reactor.Flux.MonoFlatMapMany.81445 onNext(17)
36: [tas-reader-1] - reactor.Flux.MonoFlatMapMany.81445 onNext(18)
.....

reactor.core.Exceptions$OverflowException: Queue is full: Reactive Streams source doesn't respect backpressure
	at reactor.core.Exceptions.failWithOverflow(Exceptions.java:234)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.onNext(FluxPublishOn.java:232)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:192)
	at reactor.core.publisher.FluxConcatArray$ConcatArraySubscriber.onNext(FluxConcatArray.java:176)
	at reactor.core.publisher.FluxJust$WeakScalarSubscription.request(FluxJust.java:99)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:2116)
	at reactor.core.publisher.FluxPeek$PeekSubscriber.request(FluxPeek.java:130)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runAsync(FluxPublishOn.java:446)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:526)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:1)

```

**Comment by koldat** [comment]
I have pushed the change to same PR that fixes Flux.just multithread issue.

**Comment by OlegDokuka** [comment]
@koldat indeed looks like drain loop may request more then necessary. I will be looking at that at the morning

**Comment by koldat** [comment]
@OlegDokuka yes it does. See details on what is happening in referenced PR. It basically double count missed request sometimes.

**Comment by koldat** [comment]
@OlegDokuka I have applied custom jar with my fix to our application. Before it failed after couple of seconds (20,000 requests). Right now it was running whole night and there is not a single issue after 60,000,000 requests. It is quite a good prove. Please review the fixes. I will meanwhile try to test issue that was related to rsocket 1.0.3.

**Comment by OlegDokuka** [comment]
@koldat yeah, I noticed

**Comment by simonbasle** [comment]
since there is an ongoing effort to fix this in #2576, I'll slate that for 3.3.14. However this is provisional on the review: we should be extra careful in touching the `MultiSubscriptionSubscriber` as it potentially affects other operators that extend it.

**Comment by koldat** [comment]
I understand it is critical part of sources. That is why I have added state logging of the variables which are not working properly. As a side effect we are corrupting our data streams, because we cannot rely on it (duplicating values) so for us it is serious issue. I know our application is maybe too much loaded and is heavily multithreaded, but still it should work.

Please let me know if you need more "thread recordings" or details. Tests I posted are easily to run to see the issue. In short:
1. Incorrect calculation of count to be requested when two threads hits the same MultiSubscriptionSubscriber (onComplete and request). Details in PR.
2. Flux.just does not work in multithread as two threads are using "terminado" variable that is not atomic/volatile.

After both fixes everything works perfect again.
